File: plugins/memory-vault-sync/scripts/memory_vault_runtime/privacy.py
"""Secret, local-path, and remote-document privacy enforcement."""

from __future__ import annotations

import re
import unicodedata
from typing import Any, Mapping

from memory_vault_runtime.errors import PrivacyError


MAX_VISIBLE_TEXT_BYTES = 2 * 1024 * 1024

SECRET_PATTERNS: tuple[tuple[str, re.Pattern[str]], ...] = (
    (
        "github_token",
        re.compile(r"(?:gh[pousr]_[A-Za-z0-9]{20,}|github_pat_[A-Za-z0-9_]{20,})"),
    ),
    (
        "anthropic_api_key",
        re.compile(
            r"(?<![A-Za-z0-9_-])sk-ant-api[0-9]{2}-[A-Za-z0-9_-]{20,}"
            r"(?![A-Za-z0-9_-])"
        ),
    ),
    (
        "openai_api_key",
        re.compile(
            r"(?<![A-Za-z0-9_-])sk-(?:proj-)?[A-Za-z0-9_-]{20,}"
            r"(?![A-Za-z0-9_-])"
        ),
    ),
    ("google_api_key", re.compile(r"AIza[0-9A-Za-z_-]{20,}")),
    ("aws_access_key", re.compile(r"\b(?:AKIA|ASIA)[0-9A-Z]{16}\b")),
    ("oauth_access_token", re.compile(r"\bya29\.[A-Za-z0-9_-]{20,}\b")),
    ("oauth_client_secret", re.compile(r"\bGOCSPX-[A-Za-z0-9_-]{20,}\b")),
    (
        "jwt",
        re.compile(
            r"(?<![A-Za-z0-9_-])eyJ[A-Za-z0-9_-]{8,}\."
            r"[A-Za-z0-9_-]{8,}\.[A-Za-z0-9_-]{8,}"
            r"(?![A-Za-z0-9_-])"
        ),
    ),
    (
        "bearer_token",
        re.compile(
            r"(?i)\bAuthorization\s*:\s*Bearer\s+[A-Za-z0-9._~+/=-]{12,}"
        ),
    ),
    (
        "bearer_token_value",
        re.compile(r"(?i)\bBearer\s+[A-Za-z0-9._~+/=-]{20,}"),
    ),
    ("slack_token", re.compile(r"\bx(?:ox[baprs]|app)-[A-Za-z0-9-]{10,}\b")),
    ("gitlab_token", re.compile(r"\b(?:glpat|glrt|gloas)-[A-Za-z0-9_-]{20,}\b")),
    ("npm_token", re.compile(r"\bnpm_[A-Za-z0-9]{30,}\b")),
    ("pypi_token", re.compile(r"\bpypi-[A-Za-z0-9_-]{20,}\b")),
    ("huggingface_token", re.compile(r"\bhf_[A-Za-z0-9]{30,}\b")),
    (
        "stripe_secret_key",
        re.compile(r"\b(?:sk|rk)_(?:live|test)_[A-Za-z0-9]{16,}\b"),
    ),
    (
        "sendgrid_token",
        re.compile(r"\bSG\.[A-Za-z0-9_-]{16,}\.[A-Za-z0-9_-]{16,}\b"),
    ),
    ("telegram_bot_token", re.compile(r"\b\d{8,12}:[A-Za-z0-9_-]{30,}\b")),
    ("digitalocean_token", re.compile(r"\bdop_v1_[0-9a-fA-F]{64}\b")),
    ("square_token", re.compile(r"\bsq0(?:atp|csp)-[A-Za-z0-9_-]{20,}\b")),
    ("cookie_header", re.compile(r"(?i)\b(?:Cookie|Set-Cookie)\s*:\s*\S+")),
    ("private_key", re.compile(r"-----BEGIN [A-Z0-9 ]*PRIVATE KEY-----")),
    (
        "memory_reconciliation_session_token",
        re.compile(
            r"(?i)(?:--session-token|session[_-]?token)"
            r"\s*(?:=|:|\s)\s*[\"']?[A-Za-z0-9_-]{43}"
            r"(?![A-Za-z0-9_-])"
        ),
    ),
    (
        "memory_vault_handoff_capability",
        re.compile(
            r"(?i)(?:\[\[\s*)?memory-vault-handoff\s*:\s*"
            r"[A-Za-z0-9_-]{43}(?![A-Za-z0-9_-])"
        ),
    ),
    (
        "memory_vault_routing_decision_capability",
        re.compile(r"mvrd_[A-Za-z0-9_-]{43}"),
    ),
    (
        "generic_secret_assignment",
        re.compile(
            r"(?i)\b(?:password|passwd|api[_-]?(?:key|token)|"
            r"access[_-]?token|refresh[_-]?token|oauth[_-]?token|"
            r"session[_-]?token|client[_-]?secret|secret[_-]?key|"
            r"private[_-]?key|webhook[_-]?secret)\s*[:=]\s*"
            r"[\"']?[^\s\"']{12,}"
        ),
    ),
)

ABSOLUTE_PATH_PATTERNS: tuple[re.Pattern[str], ...] = (
    re.compile(
        r"(?:^|[\s\"'(`])/(?:Users|home|root|tmp|var|private|Volumes|"
        r"content|mnt|etc|opt|Applications|Library|System|bin|sbin|"
        r"run|dev|proc|sys|srv|data|workspace|workspaces|project|projects|"
        r"repo|repos|build|app|usr/(?:local|bin|sbin|share|lib|include))/"
    ),
    re.compile(r"(?:^|[\s\"'(`])[A-Za-z]:[\\/]"),
    re.compile(r"(?:^|[\s\"'(`])\\\\[^\\\s]+\\[^\\\s]+"),
    re.compile(r"(?:^|[\s\"'(`])~[\\/]"),
)

REMOTE_FORBIDDEN_KEYS = {
    "local_path",
    "cwd",
    "session_id",
    "turn_id",
    "workspace_instance_id",
    "hostname",
    "username",
    "user",
    "email",
    "token",
    "password",
    "cookie",
    "credential",
    "upload_session_url",
}


def scan_text_content(
    value: str,
    label: str,
    *,
    reject_absolute_paths: bool = True,
) -> str:
    normalized = unicodedata.normalize("NFC", value)
    for name, pattern in SECRET_PATTERNS:
        if pattern.search(normalized):
            raise PrivacyError(f"{label} contains {name}")
    if reject_absolute_paths:
        for pattern in ABSOLUTE_PATH_PATTERNS:
            if pattern.search(normalized):
                raise PrivacyError(f"{label} contains a local absolute path")
    return normalized


def scan_visible_text(value: Any, label: str) -> str:
    if not isinstance(value, str):
        raise PrivacyError(f"{label} must be text")
    if "\x00" in value:
        raise PrivacyError(f"{label} contains NUL")
    if len(value.encode("utf-8")) > MAX_VISIBLE_TEXT_BYTES:
        raise PrivacyError(f"{label} is too large")
    return scan_text_content(value, label)


def assert_remote_safe(value: Any, trail: tuple[str, ...] = ()) -> None:
    if isinstance(value, Mapping):
        for key, child in value.items():
            lowered = str(key).lower()
            if lowered in REMOTE_FORBIDDEN_KEYS:
                raise PrivacyError(
                    "remote document contains forbidden key: "
                    f"{'.'.join((*trail, lowered))}"
                )
            assert_remote_safe(child, (*trail, str(key)))
    elif isinstance(value, list):
        for child in value:
            assert_remote_safe(child, trail)
    elif isinstance(value, str):
        scan_visible_text(value, "remote text")

File: clients/typescript/network/privacy.ts
/** The existing finite publication heuristic, used only by guarded Hint exports.
 * This is not comprehensive DLP. It does not alter local memory or grant access.
 * Pattern sources mirror memory_vault_privacy.py; errors never echo matches.
 */
import { NetworkCryptoError } from './crypto.ts';

const MAX_SCAN_BYTES = 128 * 1024 * 1024, MAX_SCAN_RECORDS = 100000;
function fail(code:string):never { throw new NetworkCryptoError(code); }
// Python's Unicode word/space classes differ from JavaScript's ASCII \\b and
// \\w. Spell out those classes when compiling the shared finite patterns.
const WORD = String.raw`[\p{L}\p{N}_]`;
const BOUNDARY = `(?:(?<!${WORD})(?=${WORD})|(?<=${WORD})(?!${WORD}))`;
const SPACE = String.raw`[\p{White_Space}\u001c-\u001f]`;
function pattern(source:string,insensitive:boolean):RegExp {
  source=source.replace('(?i)','');
  if(insensitive){
    // Python IGNORECASE also folds dotted/dotless I into ASCII i. JS /iu
    // already handles long s and Kelvin sign, but omits these two forms.
    let folded='',position=0;
    while(position<source.length){
      const character=source[position++];
      if(character==='\\'){folded+=character+(source[position++]??'');continue;}
      if(character==='['){
        let field='';
        while(position<source.length){const next=source[position++];if(next==='\\'){field+=next+(source[position++]??'');continue;}if(next===']')break;field+=next;}
        folded+='['+field+(/[iI]|A-Z|a-z/.test(field)?String.raw`\u0130\u0131`:'')+']';continue;
      }
      folded+=character==='i'||character==='I'?String.raw`[iI\u0130\u0131]`:character;
    }
    source=folded;
  }
  let adapted='',inClass=false;
  for(let position=0;position<source.length;position++){
    const character=source[position];
    if(character==='\\'){
      const token=source[++position];
      adapted+=token==='b'?BOUNDARY:token==='d'?String.raw`\p{Nd}`:
        token==='s'?(inClass?String.raw`\p{White_Space}\u001c-\u001f`:SPACE):
        token==='S'?String.raw`[^\p{White_Space}\u001c-\u001f]`:
        token==='"'||token==="'"?token:'\\'+token;
    }else{if(character==='[')inClass=true;if(character===']')inClass=false;adapted+=character;}
  }
  return new RegExp(adapted,insensitive?'iu':'u');
}
const SECRETS = [
  pattern("-----BEGIN [A-Z0-9 ]*PRIVATE KEY-----",false),
  pattern("(?:gh[pousr]_[A-Za-z0-9]{20,}|github_pat_[A-Za-z0-9_]{20,})",false),
  pattern("(?<![A-Za-z0-9_])sk-(?:proj-|svcacct-)?[A-Za-z0-9_-]{20,}(?![A-Za-z0-9_-])",false),
  pattern("\\b(?:AKIA|ASIA)[A-Z0-9]{16}\\b",false),
  pattern("AIza[A-Za-z0-9_-]{20,}",false),
  pattern("\\b(?:ya29\\.|GOCSPX-)[A-Za-z0-9_-]{20,}\\b",false),
  pattern("(?<![A-Za-z0-9_-])eyJ[A-Za-z0-9_-]{8,}\\.[A-Za-z0-9_-]{8,}\\.[A-Za-z0-9_-]{8,}(?![A-Za-z0-9_-])",false),
  pattern("(?i)\\bauthorization\\s*:\\s*(?:bearer|basic)\\s+[A-Za-z0-9._~+/=-]{12,}",true),
  pattern("(?i)\\bBearer\\s+[A-Za-z0-9._~+/=-]{20,}",true),
  pattern("\\bx(?:ox[baprs]|app)-[A-Za-z0-9-]{10,}\\b",false),
  pattern("\\b(?:glpat|glrt|gloas)-[A-Za-z0-9_-]{20,}\\b",false),
  pattern("\\b(?:npm_|hf_)[A-Za-z0-9]{30,}\\b",false),
  pattern("\\b(?:pypi-|sq0(?:atp|csp)-)[A-Za-z0-9_-]{20,}\\b",false),
  pattern("\\b(?:sk|rk)_(?:live|test)_[A-Za-z0-9]{16,}\\b",false),
  pattern("\\bSG\\.[A-Za-z0-9_-]{16,}\\.[A-Za-z0-9_-]{16,}\\b",false),
  pattern("\\b\\d{8,12}:[A-Za-z0-9_-]{30,}\\b",false),
  pattern("\\bdop_v1_[0-9a-fA-F]{64}\\b",false),
  pattern("(?i)\\b(?:Cookie|Set-Cookie)\\s*:\\s*\\S+",true),
  pattern("(?i)(?:--session-token|session[_-]?token)(?:\\s*[:=]\\s*|\\s+)[\\\"']?[A-Za-z0-9_-]{43}(?![A-Za-z0-9_-])",true),
  pattern("(?i)(?:\\[\\[\\s*)?memory-vault-handoff\\s*:\\s*[A-Za-z0-9_-]{43}(?![A-Za-z0-9_-])",true),
  pattern("mvrd_[A-Za-z0-9_-]{43}",false),
  pattern("(?i)\\b(?:password|passwd|api[_-]?(?:key|token)|access[_-]?token|refresh[_-]?token|oauth[_-]?token|session[_-]?token|client[_-]?secret|secret[_-]?key|private[_-]?key|webhook[_-]?secret)\\s*[:=]\\s*[\\\"']?[^\\s\\\"']{12,}",true),
  pattern("\\bhttps?://[^\\s/@:]{1,128}:[^\\s/@]{1,256}@",false),
];
const LOCAL_PATHS = [
  pattern("(?<![A-Za-z0-9])/(?:Users|home|private|var|tmp|Volumes)/[^\\s\\\"<>]+",false),
  pattern("(?i)\\b[A-Z]:[\\\\/](?:Users|Documents and Settings)[\\\\/][^\\r\\n\\\"<>]+",true),
  pattern("(?:^|[\\s\\\"'(`])/(?:Users|home|root|tmp|var|private|Volumes|content|mnt|etc|opt|Applications|Library|System|bin|sbin|run|dev|proc|sys|srv|data|workspace|workspaces|project|projects|repo|repos|build|app|usr/(?:local|bin|sbin|share|lib|include))/",false),
  pattern("(?:^|[\\s\\\"'(`])[A-Za-z]:[\\\\/]",false),
  pattern("(?:^|[\\s\\\"'(`])\\\\\\\\[^\\\\\\s]+\\\\[^\\\\\\s]+",false),
  pattern("(?:^|[\\s\\\"'(`])~[\\\\/]",false),
];

/** Match the Python scanner's raw + NFC string traversal and finite budgets.
 * Inputs here are validated immutable records, not executable object content.
 */
export function assertPublishable(records:readonly Record<string,unknown>[]):void {
  let remaining=MAX_SCAN_BYTES,nodes=MAX_SCAN_BYTES,count=0;
  for(const record of records){
    if(++count>MAX_SCAN_RECORDS||record===null||typeof record!=='object'||Array.isArray(record))fail('publication_scan_limit');
    let secret=false,local=false;
    const pending:{iterator:Iterator<unknown>;depth:number}[]=[{iterator:[record][Symbol.iterator](),depth:0}];
    while(pending.length){
      const frame=pending[pending.length-1],next=frame.iterator.next();
      if(next.done){pending.pop();continue;}
      if(frame.depth>32||--nodes<0)fail('publication_scan_limit');
      const value=next.value;
      if(Array.isArray(value)){pending.push({iterator:value[Symbol.iterator](),depth:frame.depth+1});}
      else if(value!==null&&typeof value==='object'){
        function* parts():IterableIterator<unknown>{for(const [key,child] of Object.entries(value)){yield key;yield child;}}
        pending.push({iterator:parts(),depth:frame.depth+1});
      }else if(typeof value==='string'){
        const raw=Buffer.from(value,'utf8');if(raw.toString('utf8')!==value)fail('publication_invalid_text');
        remaining-=raw.length;if(remaining<0)fail('publication_scan_limit');
        const normalized=value.normalize('NFC'),projections=normalized===value?[value]:[value,normalized];
        for(const text of projections){
          secret ||= SECRETS.some(expression=>expression.test(text));
          local ||= LOCAL_PATHS.some(expression=>expression.test(text));
        }
      }
    }
    if(secret)fail('publication_secret_detected');
    if(local)fail('publication_local_path_detected');
  }
}
